src/main/java/io/cryostat/net/web/http/api/v2/RuleDeleteHandler.java

```diff
@@ -99,16 +121,68 @@ public HttpMimeType mimeType() {
         }
+        Rule rule = ruleRegistry.getRule(name).get();
         try {
-            ruleRegistry.deleteRule(name);
+            ruleRegistry.deleteRule(rule);
         } catch (IOException e) {
             throw new ApiException(500, "IOException occurred while deleting rule", e);
         }
-        return new IntermediateResponse<Void>().body(null);
+        List<CleanupFailure> failures = new ArrayList<>();
+        if (Boolean.valueOf(params.getQueryParams().get(CLEAN_PARAM))) {
+            for (ServiceRef ref : platformClient.listDiscoverableServices()) {
+                if (!ruleRegistry.applies(rule, ref)) {
+                    continue;
+                }
+                try {
+                    targetConnectionManager.executeConnectedTask(
+                            new ConnectionDescriptor(ref, credentialsManager.getCredentials(ref)),
+                            conn -> {
+                                conn.getService().getAvailableRecordings().stream()
+                                        .filter(
+                                                rec ->
+                                                        rec.getName()
+                                                                .equals(rule.getRecordingName()))
+                                        .findFirst()
+                                        .ifPresent(
+                                                r -> {
+                                                    try {
+                                                        conn.getService().stop(r);
+                                                    } catch (Exception e) {
+                                                        logger.error(new ApiException(500, e));
+                                                        CleanupFailure failure =
+                                                                new CleanupFailure();
+                                                        failure.ref = ref;
+                                                        failure.message = e.getMessage();
+                                                        failures.add(failure);
+                                                    }
+                                                });
+                                return null;
+                            });
+                } catch (Exception e) {
+                    logger.error(new ApiException(500, e));
+                    CleanupFailure failure = new CleanupFailure();
+                    failure.ref = ref;
+                    failure.message = e.getMessage();
+                    failures.add(failure);
+                }
+            }
+        }
+        if (failures.size() == 0) {
+            return new IntermediateResponse<List<RuleDeleteHandler.CleanupFailure>>().body(null);
+        } else {
+            return new IntermediateResponse<List<CleanupFailure>>().statusCode(500).body(failures);
```

Comment:
What is the difference between `List<RuleDeleteHandler.CleanupFailure>` and `List<CleanupFailure>` in this context?

Reply:
Ah, nothing really. They refer to the same type, but I suppose my editor/code-completion did it one way and I hand-wrote it the other. I'll clean that up.